src/main/config.cpp

```diff
@@ -0,0 +1,99 @@
+#include <cstdio>
+#include <iostream>
+#include <sstream>
+#include <gflags/gflags.h>
+#include "config.hpp"
+
+const char DELIMITER = ',';
+
+/*
+	FLAGS SECTION
+*/
+
+DEFINE_string(monitor, "exec", "Defines the usecase of the program. The default value is \"exec\", which monitors process executions.");
+DEFINE_validator(monitor, Config::check_monitor_type);
+
+// EXEC FLAG - MONITOR PROCESS EXECUTION
+DEFINE_string(ppid, "", "Specifies the parent pid(s) of the processes to monitor using comma-separated values.");
+DEFINE_string(name, "", "Specifies the name(s) of the processes to monitor using comma-separated values.");
+DEFINE_int32(n, 0, "Specifies the number of processes to monitor. The program will stop after n processes were executed.");
+
+// DATA EXPORT FLAGS
+DEFINE_string(format, "", "Specifies the format in which to save the terminal output in a separate file. Available formats are csv and protobuf.");
+DEFINE_validator(format, Config::check_format_type);
+DEFINE_string(output_file, "", "Specifies the path to write the ouput or where to create the file, if it does not exist at that path.");
+
+/*
+	END OF FLAGS SECTION
+*/
+
+std::map<std::string, Usecase> usecase_map = {
+    {"exec", EXEC_MONITOR}
+};
+
+Config::Config() {}
+
+int Config::parse_flags(int argc, char *argv[]) {
```

Comment:
I already suggested a few times that the initialization of the Config object should be done in the constructor of the Config class. If your object can end up in an invalid state, I suggest this pseudo-code:

```
Config::Config() {
	gflags::ParseCommandLineFlags(&argc, &argv, true);
	int ret = 0;

        valid = true;
	switch (usecase_map[FLAGS_monitor]) {
	case Usecase::EXEC_MONITOR:
		usecase = EXEC_MONITOR;
		ret = exec_monitor_parse_args();
		break;
	default:
		valid = false;
	}

	gflags::ShutDownCommandLineFlags();
}

bool Config::isValid() {
    return valid;
}
```

```
	Config config;
	if (!config.isValid()) {
                std::cerr << "Incorrect use. Run hawk --help for instructions.\n";
		return -1;
	}
```

The reason why I keep insisting on this is that an object initialization should be done in the constructor. This is a basic principle of object oriented design that once you called the constructor the object is in a state that makes it kinda usable. Also in this particular case, this would enforce that the object is initialized only once because with your code, nothing enforces that the user doesn't call parse_flags() twice, maybe with different parameters and then get an object in an inconsistent state

Reply:
Oh, thank you for pointing out again. I modified according to your suggestions, and I have one observation: it is not necessary to validate the value for the monitor flag in the switch, as it is already done via the validator set from gflags. So, now, the is_valid variable only checks if the rest of the flags are correctly given, and if not, stops the execution.